docs/specification/2025-03-26/basic/authorization.mdx

```diff
@@ -8,303 +8,497 @@ title: Authorization
-authorization flow after receiving the _HTTP 401 Unauthorized_.
+MCP servers **MUST** follow [RFC 6750 Section 3.1](https://www.rfc-editor.org/rfc/rfc6750#section-3.1) for
+any authorization failures.
 
-The following demonstrates the basic OAuth 2.1 for public clients using PKCE.
+For MCP server capability discovery:
+
+- MCP clients **MUST** parse the `WWW-Authenticate` response to discover the path to the PRM document.
+- MCP servers **SHOULD** include the PRM document path in error responses.
+- MCP clients **SHOULD** evaluate or enable the user to evaluate the trustworthiness of authorization
+  servers they are connecting to.
+- Authorization servers **MUST** support
+  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)
+
+The discovery flow is illustrated below:  
 
 ```mermaid
 sequenceDiagram
-    participant B as User-Agent (Browser)
-    participant C as Client
-    participant M as MCP Server
-
-    C->>M: MCP Request
-    M->>C: HTTP 401 Unauthorized
-    Note over C: Generate code_verifier and code_challenge
-    C->>B: Open browser with authorization URL + code_challenge
-    B->>M: GET /authorize
-    Note over M: User logs in and authorizes
-    M->>B: Redirect to callback URL with auth code
-    B->>C: Callback with authorization code
-    C->>M: Token Request with code + code_verifier
-    M->>C: Access Token (+ Refresh Token)
-    C->>M: MCP Request with Access Token
-    Note over C,M: Begin standard MCP message exchange
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant AS as Authorization Server
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+    MC->>MS: Fetch PRM document
+    MS-->>MC: PRM document response
+    Note over MC: Validate PRM document<br/>metadata
+    Note over MC: Build AS metadata URL
+    MC->>AS: Fetch AS metadata
+    AS-->>MC: AS metadata response
+
+    Note over MC,AS: Continue with authorization flow
+```
+
+#### 2.3.1 Server Metadata Discovery Headers
+
+MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
+Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.
+
+For example:
+
+```http
+MCP-Protocol-Version: 2024-11-05
 ```
 
-### 2.3 Server Metadata Discovery
+### 2.4 Authorization server integration
 
-For server capability discovery:
+#### 2.4.1 Overview
 
-- MCP clients _MUST_ follow the OAuth 2.0 Authorization Server Metadata protocol defined
-  in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).
-- MCP server _SHOULD_ follow the OAuth 2.0 Authorization Server Metadata protocol.
-- MCP servers that do not support the OAuth 2.0 Authorization Server Metadata protocol,
-  _MUST_ support fallback URLs.
+MCP servers **MAY** override endpoints on an as-needed basis if the authorization server requires
```

Comment:
This paragraph should be removed. We want to ensure there is only one way for the client to discover the authorization server. The AS can be colocated with an RS if the server operator wishes to deploy it in that way, but the discovery mechanism would still be the same as if they were separate.

Reply:
Updated.